Dart/src/com/jetbrains/lang/dart/ide/completion/DartServerCompletionContributor.java

```diff
@@ -455,8 +479,46 @@ public static LookupElementBuilder createLookupElement(@NotNull final Project pr
       }
     }
 
-    // Use selection offset / length.
-    if (shouldSetSelection) {
+    if (isNotYetImported) {
+      lookup = lookup.withInsertHandler((context, item) -> {
+        final DartAnalysisServerService das = DartAnalysisServerService.getInstance(project);
+        final GetCompletionDetailsResult result = das.completion_getSuggestionDetails(file, suggestionSetId, suggestion.getCompletion(),
+                                                                                      context.getStartOffset());
+        if (result == null) {
+          return;
+        }
+
+        context.getDocument().replaceString(context.getStartOffset(), context.getTailOffset(), result.completion);
```

Comment:
Why do we need this line? Isn't it effectively a noop?

Reply:
> > TODO: Either in this CL or a follow-up we need to update the client to respect relevance tags.
> 
> You already use it, don't you? Or what do you mean?

That's client support for https://github.com/dart-lang/sdk/commit/b9bcfb4529905a2533200ad11547db67740fdc6c.

Reply:
Here the client asks server for the full replacement since an import may be needed in addition to the cursor-position completion insertion. This completion might not be the same as what we put in the LookupElement. Does that make sense?

Reply:
> This completion might not be the same as what we put in the LookupElement. Does that make sense?

If so then yes. Great scenario for a test!

Reply:
I've updated the test a bit so that it shows `Process` but inserts `prefix0.Process`. Without this line it fails, so this line is now covered with tests.